firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java

```diff
@@ -329,13 +335,42 @@ public WriteBatch batch() {
     return batch.commit();
   }
 
-  @VisibleForTesting
-  Task<Void> shutdown() {
+  Task<Void> shutdown(boolean fromAppDeletion) {
+    if (!fromAppDeletion && this.getApp() != null) {
+      FirestoreMultiDbComponent component = this.getApp().get(FirestoreMultiDbComponent.class);
```

Comment:
You should not modify the errorprone check for "convenience" as this convenience comes at a price:

* You're essentially gaining access to any *undeclared* component dependency. This is something we(acore) want teams to do only if absolutely necessary and only as a temporary measure that they should consider tech debt.
* It makes the code harder to unit test as you're no longer getting everything you need via a constructor but effectively gaining access to services through another object you have access to via a "service locator" pattern(see https://github.com/google/guice/wiki/InjectOnlyDirectDependencies)
* Additionally, in this particular case, you are coupling this class to `FirestoreMultiDbComponent` imo unnecessarily

That said, I'd recommend something along the following lines:

* add a constructor parameter to `FirebaseFirestore` in the form of `Consumer<String> onShutdown`(or equivalent as it is java8 only)
* have `FirestoreMultiDbComponent` inject the `Consumer<String>` into `Firestore` as a simple lambda that removes the instance based on the provided `String` `dbId`
* have firestore call `onShutdown` when appropriate.

Reply:
Basically what you suggested, slight tweak.